persistent/Database/Persist/Quasi/Internal/ModelParser.hs

```diff
@@ -0,0 +1,536 @@
+equality = label "equality expression" $ do
+    L.lexeme spaceConsumer $ do
+        lhs <- some contentChar
+        _ <- char '='
+        rhs <-
+            choice
+                [ try quotation'
+                , try sqlLiteral
+                , try parentheticalInner
+                , some $ contentChar <|> char '(' <|> char ')'
+                ]
+        pure $ Equality (pack lhs) (pack rhs)
+  where
+    parentheticalInner = do
+        str <- parenthetical'
+        pure . init . drop 1 $ str
+
+sqlTypeName :: Parser String
+sqlTypeName = some $ choice [ alphaNumChar
+                            , char '_'
+                            ]
+
+sqlLiteral :: Parser String
+sqlLiteral = label "SQL literal" $ do
+  quote <- L.lexeme spaceConsumer $ char '\'' *> manyTill charLiteral (char '\'')
+  st <- optional $ do
+    colons <- string "::"
+    tn <- sqlTypeName
+    pure $ colons <> tn
+  pure $ mconcat [ "'"
+                 , quote
+                 , "'"
+                 , maybe "" id st
+                 ]
+
+quotation :: Parser Token
+quotation = label "quotation" $ do
+    str <- L.lexeme spaceConsumer quotation'
+    pure . Quotation $ pack str
+
+quotation' :: Parser String
+quotation' = char '"' *> manyTill charLiteral (char '"')
+
+parenthetical :: Parser Token
+parenthetical = label "parenthetical" $ do
+    str <- L.lexeme spaceConsumer parenthetical'
+    pure . Parenthetical . pack . init . drop 1 $ str
+
+parenthetical' :: Parser String
+parenthetical' = do
+    str <- between (char '(') (char ')') q
+    pure $ "(" ++ str ++ ")"
+  where
+    q = mconcat <$> some (c <|> parenthetical')
+    c = (: []) <$> choice [contentChar, nonLineSpaceChar, char '"']
+
+blockKey :: Parser Token
+blockKey = label "block key" $ do
+    fl <- upperChar
+    rl <- many alphaNumChar
+    pure . BlockKey . pack $ fl : rl
+
+ptext :: Parser Token
+ptext = label "plain token" $ do
+    str <- L.lexeme spaceConsumer $ some contentChar
+    pure . PText . pack $ str
+
+docComment :: Parser Token
+docComment = label "doc comment" $ do
+    _ <- string "-- |" <* hspace
+    str <- many (contentChar <|> nonLineSpaceChar)
+    pure . DocComment . pack $ str
+
+comment :: Parser Token
+comment = label "comment" $ do
+    _ <- (string "--" <|> string "#") <* hspace
+    str <- many (contentChar <|> nonLineSpaceChar)
+    pure . Comment . pack $ str
+
+anyToken :: Parser Token
+anyToken =
+    choice
+        [ try equality
+        , docComment
+        , comment
+        , quotation
+        , parenthetical
+        , ptext
+        ]
+
+data ParsedEntityDef = ParsedEntityDef
+    { parsedEntityDefComments :: [Text]
+    , parsedEntityDefEntityName :: EntityNameHS
+    , parsedEntityDefIsSum :: Bool
+    , parsedEntityDefEntityAttributes :: [Attr]
+    , parsedEntityDefFieldAttributes :: [[Token]]
+    , parsedEntityDefFieldComments :: [Maybe Text]
+    , parsedEntityDefExtras :: M.Map Text [ExtraLine]
+    , parsedEntityDefSpan :: Maybe Span
```

Comment:
y'know while we're at it, we may want to, uh, change the name of the `Span` type - wish I'd caught this earlier. maybe `SourceSpan` or something?

Reply:
Yeah, that would be fantastic